### PR TITLE
fix: Contains fix for items getting filtered in merchant field : https://app.clickup.com/t/85ztztwg6

### PR DESCRIPTION
**Pull request by howdysuraj**
### <samp>🤖 Generated by Copilot at e0640b8</samp>

This pull request enhances the `VirtualSelectComponent` by using interfaces to define its inputs and outputs, and by fixing some display issues for the merchant name. It also introduces the `virtual-select.model.ts` file, which contains the interface definitions.

### <samp>🤖 Generated by Copilot at e0640b8</samp>

> _`VirtualSelect`_
> _Types and interfaces / kireji_
> _Autumn of refactor_

### <samp>🤖 Generated by Copilot at e0640b8</samp>

*  Define interfaces for virtual select data in `virtual-select.model.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-d9a465bcbe0054810f51022c33a11a239c006369fb6f3cadcc98ae81bc4ec3a0R1-R21))
*  Import interfaces from `virtual-select.model.ts` in `virtual-select.component.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dR8))
*  Replace `any` type with `Partial<VirtualSelectOptionValues>` for `options` input, `value` property and getter/setter, and `writeValue` method parameter in `virtual-select.component.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL22-R23), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL46-R53), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL74-R80), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL128-R130), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL135-R144))
*  Replace `any` type with `unknown` for `selectionElement` input in `virtual-select.component.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL30-R31))
*  Add type annotations for `defaultLabelProp`, `recentlyUsed`, `displayValue`, `innerValue`, `valid` getter, `ngOnInit` and `openModal` methods in `virtual-select.component.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL46-R53), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL66-R68), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL93-R104))
*  Add eslint comments to disable rules for intentional use of `any` type in `registerOnChange`, `registerOnTouched`, `openModal` and `writeValue` methods in `virtual-select.component.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dR59), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dR90), [link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL150-R172))
*  Add exception for `display_name` property in `writeValue` method to display merchant name in `virtual-select.component.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2539/files?diff=unified&w=0#diff-17229282389aab9699893e3b52b4b331d1a64de11ef4d080fa16866eadc4ff7dL150-R172))

## Clickup
https://app.clickup.com/t/85ztztwg6

## Code Coverage
Please add code coverage here

## UI Preview
localhost_8100_enterprise_add_edit_expense;id=txeg8fCtVgj8;persist_filters=true(iPhone 12 Pro)

Screenshot 2023-11-06 at 9 29 25 AM

Although this is a product issue this hack has been added to sync the behaviour of web and mobile app

### CODE REVIEW

**Comment by howdysuraj** [comment]
Screenshot 2023-11-06 at 8 40 43 AM
A test is failing in Add Corporate card component, what should we do for this ?

File: src/app/shared/components/virtual-select/virtual-select.component.ts [hunk 1]
```diff
@@ -19,14 +20,15 @@ import { ModalPropertiesService } from 'src/app/core/services/modal-properties.s
...
-  @Input() options: { label: string; value: any }[] = [];
+  @Input() options: { label: string; value: VirtualSelectOptions }[] = [];
 
   @Input() disabled = false;
 
   @Input() label = '';
 
   @Input() mandatory = false;
 
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (OmkarJ13) [inline src/app/shared/components/virtual-select/virtual-select.component.ts:31]: Could we avoid this? Let's type TemplateRef to something other than any

File: src/app/shared/components/virtual-select/virtual-select.component.ts [hunk 2]
```diff
@@ -43,18 +45,19 @@ export class VirtualSelectComponent implements ControlValueAccessor, OnInit {
...
 
-  private innerValue;
+  private innerValue: VirtualSelectOptions;
 
   private ngControl: NgControl;
 
   private onTouchedCallback: () => void = noop;
 
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
   private onChangeCallback: (_: any) => void = noop;
```
Comment (OmkarJ13) [inline src/app/shared/components/virtual-select/virtual-select.component.ts:61]: Same as above, let's define the parameter type for the function, in all the places

File: src/app/shared/components/virtual-select/virtual-select.component.ts [hunk 3]
```diff
@@ -85,21 +88,21 @@ export class VirtualSelectComponent implements ControlValueAccessor, OnInit {
         } else if (typeof this.innerValue === 'string') {
           this.displayValue = this.innerValue;
         } else if (this.innerValue && this.defaultLabelProp) {
+          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
           this.displayValue = this.innerValue[this.defaultLabelProp];
```
Comment (OmkarJ13) [inline src/app/shared/components/virtual-select/virtual-select.component.ts:92]: No other way than to disable the rule?

File: src/app/shared/components/virtual-select/virtual-select.component.ts [hunk 4]
```diff
@@ -125,41 +128,57 @@ export class VirtualSelectComponent implements ControlValueAccessor, OnInit {
...
+        this.handleDisplayNameException();
+      }
+    }
+  }
+
+  //Hack for display_name values added to resolve BR related to merchant name not in list, BR: https://app.clickup.com/t/85ztztwg6;
+  handleDisplayNameException(): void {
+    if (this.innerValue && this.innerValue.hasOwnProperty('display_name') && !this.displayValue) {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/dot-notation
+      const display_name = this.innerValue['display_name'];
```
Comment (OmkarJ13) [inline src/app/shared/components/virtual-select/virtual-select.component.ts:165]: displayName

**Review by OmkarJ13 — COMMENTED** [review]
Actual logic LGTM, left some comments on the linting part.

Want to avoid disabling lint rules unless absolutely necessary

File: src/app/shared/components/virtual-select/virtual-select.component.ts (hunk 1 above)
Comment (howdysuraj) [reply src/app/shared/components/virtual-select/virtual-select.component.ts:31]: @OmkarJ13 its ok to use any for template ref, previously used unknown , dhar recommended it ok to use any for Template ref

File: src/app/shared/components/virtual-select/virtual-select.component.ts (hunk 4 above)
Comment (howdysuraj) [reply src/app/shared/components/virtual-select/virtual-select.component.ts]: fixed

File: src/app/shared/components/virtual-select/virtual-select.component.ts (hunk 3 above)
Comment (howdysuraj) [reply src/app/shared/components/virtual-select/virtual-select.component.ts:91]: The whole component is designed is keeping `any` in mind so its difficult at few places or a code restructuring is required.

File: src/app/shared/components/virtual-select/virtual-select.component.ts (hunk 2 above)
Comment (howdysuraj) [reply src/app/shared/components/virtual-select/virtual-select.component.ts]: fixed

**Comment by howdysuraj** [comment]
> Actual logic LGTM, left some comments on the linting part.
> 
> Want to avoid disabling lint rules unless absolutely necessary

The component was designed in keeping `any` in mind , so its difficult to get rid of any at some places, wherever possible types are added.

**Review by OmkarJ13 — APPROVED** [review]

**Review by Julias0 — APPROVED** [review]

**Pull request merged**